pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java

```diff
@@ -237,6 +237,21 @@ private void manageSegmentLineageCleanupForTable(TableConfig tableConfig) {
             // entry and its segments
             Set<String> destinationSegments = new HashSet<>(lineageEntry.getSegmentsTo());
             destinationSegments.retainAll(segmentsForTable);
+
+            // Retain zombie segments for the case of rerunning the replace protocol with overlapping segment names.
+            // For the following example, we should not try to delete s1 and s2, but remove entry1 directly.
+            // entry1: { segmentsFrom: [], segmentsTo: [s1, s2], status: REVERTED}
+            // entry2: { segmentsFrom: [], segmentsTo: [s1, s2], status: IN_PROGRESS/COMPLETED}
+            if (lineageEntry.getState() == LineageEntryState.REVERTED) {
+              for (String otherLineageEntryId : segmentLineage.getLineageEntryIds()) {
+                LineageEntry otherLineageEntry = segmentLineage.getLineageEntry(otherLineageEntryId);
+                if (otherLineageEntry.getState() == LineageEntryState.COMPLETED
+                    || otherLineageEntry.getState() == LineageEntryState.IN_PROGRESS) {
+                  destinationSegments.removeAll(otherLineageEntry.getSegmentsTo());
+                }
+              }
+            }
+
```

Comment:
maybe update the comment for the `if (destinationSegments.isEmpty()) {` below that empty destinationSegments might also come from entry reverted by a newer entry with overlapping segment names.

Reply:
Updated the comment.